File: spring-rest-configuration-autodiscover.java/src/main/java/demo/service/core/AbstractPropertyService.java
package demo.service.core;

import demo.domain.Property;
import demo.repository.PropertyRepository;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

abstract public class AbstractPropertyService<T> implements PropertyService<T> {

    private Optional<T> valueOpt = Optional.empty();

    @Autowired
    private PropertyRepository propertyRepository;

    @Override
    public Property update(String value) {
        valueOpt = Optional.of(fromString(value));
        return save(value);
    }

    @Override
    public void init() {
        Optional<Property> propertyOpt = findPropertyByCode(getCode());
        if (propertyOpt.isPresent()) {
            String value = propertyOpt.get().getValue();
            valueOpt = Optional.of(fromString(value));
        } else {
            T value = getDefaultValue();
            valueOpt = Optional.of(value);
            save(toString(value));
        }
    }

    public T get() {
        if (valueOpt.isPresent()) {
            return valueOpt.get();
        } else {
            Optional<Property> propertyOpt = findPropertyByCode(getCode());
            if (propertyOpt.isPresent()) {
                valueOpt = Optional.of(fromString(propertyOpt.get().getValue()));
                return valueOpt.get();
            } else {
                return getDefaultValue();
            }
        }
    }

    protected abstract Type getType();

    protected abstract T fromString(String input);

    protected abstract String toString(T input);

    protected abstract T getDefaultValue();

    public List<String> getPossibleValues() {
        return new ArrayList<>();
    }

    private Optional<Property> findPropertyByCode(Code code) {
        return propertyRepository.findByCode(code);
    }

    private Property save(String value) {
        Property property = new Property();
        property.setCode(getCode());
        property.setType(getType());
        property.setValue(value);

        return propertyRepository.save(property);
    }

    @Override
    public String toString() {
        return super.toString() + "{" +
                "getCode()=" + getCode() +
                ", getType()=" + getType() +
                '}';
    }
}

File: spring-rest-configuration-autodiscover.scala/src/main/scala/demo/service/core/AbstractPropertyService.scala
package demo.service.core

import demo.domain.Property
import demo.repository.PropertyRepository
import org.springframework.beans.factory.annotation.Autowired
import java.util._

abstract class AbstractPropertyService[T] extends PropertyService[T] {

    private var valueOpt = Optional.empty[T]()

    @Autowired
    private val propertyRepository: PropertyRepository = _

    override def update(value: String): Property = {
        valueOpt = Optional.of(fromString(value))
        save(value)
    }

    override def init() = {
        val propertyOpt = findPropertyByCode(getCode())
        if (propertyOpt.isPresent) {
            val value = propertyOpt.get().value
            valueOpt = Optional.of(fromString(value))
        } else {
            val value = getDefaultValue()
            valueOpt = Optional.of(value)
            save(toString(value))
        }
    }

    def get(): T = {
        if (valueOpt.isPresent) {
            valueOpt.get()
        } else {
            val propertyOpt = findPropertyByCode(getCode())
            if (propertyOpt.isPresent) {
                valueOpt = Optional.of(fromString(propertyOpt.get().value))
                valueOpt.get()
            } else {
                getDefaultValue()
            }
        }
    }

    protected abstract def getType: Type.Value

    protected abstract def fromString(input: String): T

    protected abstract def toString(input: T): String

    protected abstract def getDefaultValue: T

    override def getPossibleValues: java.util.List[String] = new java.util.ArrayList()

    private def findPropertyByCode(code: Code.Value): Optional[Property] = {
        propertyRepository.findByCode(code)
    }

    private def save(value: String): Property = {
        val property = new Property()
        property.code = getCode
        property.type = getType
        property.value = value

        propertyRepository.save(property)
    }

    override def toString: String = {
        super.toString + "{" +
                "getCode()=" + getCode +
                ", getType()=" + getType +
                '}'
    }
}
